Docs/lsm 482 partial compile guidelines

sanderr opened the pull request:
# Description

Added modeling guidelines for safe use of partial compiles. I decided to put it in core rather than lsm because it could pretty much all be generalized and it all applies to core-only use of partial compiles as well.

Initial technical review. In the next stage I'll see if Pouriya is available for a review on the quality of the documentation.

- I had to move some files and GitHub does not show the diff properly. Review is probably easiest with `git diff master --color-moved` or `git diff master --find-renames=5%`

closes inmanta/lsm#482

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [ ] Attached issue to pull request
- [ ] Changelog entry
- [ ] Type annotations are present
- [ ] Code is clear and sufficiently documented
- [ ] No (preventable) type errors (check using make mypy or make mypy-diff)
- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

wouterdb commented on docs/resource_sets.rst:
```diff
@@ -91,3 +93,168 @@ Limitations
+TODO: "On a high level" -> remove?
+TODO: should I simplify and just talk about resources rather than generic nodes? Less accurate but also less abstract
+
+On a high level, to allow safe use of partial compiles, it is imperative that each service's refinements (through
+implementations) form a tree that can only intersect between service instances on shared nodes. The whole subtree below such a
+shared node should be considered shared and any resources in it must not be part of a resource set. All shared resources should
+be consistent between any two service instances that might create the object (see `Constraints and rules`_). All other nodes
+should generally be considered owned by the service and all their resources be part of the service's resource set. For more
+details on what it means to own a a child node in the tree (e.g. a resource) and how to ensure two service instance's trees can
+not intersect on owned nodes, see the `Ownership`_ subsection.
```
Would it not be easier to explain from the notion of ownership?

```suggestion
On a high level, to allow safe use of partial compiles, each service must be the sole owner if its resources and shared resources must be identical across service instances. More precisely: each service's refinements (through
implementations) form a tree that can only intersect between service instances on shared nodes. The whole subtree below such a shared node should be considered shared and any resources in it must not be part of a resource set. All shared resources should be consistent between any two service instances that might create the object (see `Constraints and rules`_). All other nodes should generally be considered owned by the service and all their resources be part of the service's resource set. For more details on what it means to own a a child node in the tree (e.g. a resource) and how to ensure two service instance's trees can not intersect on owned nodes, see the `Ownership`_ subsection.
```

wouterdb commented on docs/resource_sets.rst:
```diff
@@ -91,3 +93,168 @@ Limitations
+other is the use of some external distributor of unique values (e.g. a plugin to generate a UUID or to allocate values in an
+inventory).
+
+In either case, the goal is to make sure that any object that is marked as owned by a service instance, is unique to that
```
1. The index here is in fact not required, the condition below it is sufficient:
`we do so by making sure the object's identity is in fact completely derived from the identity of the service instance` 
The index helps verifying this behavior when doing a full compile, but it not required.

2. completely derived is not sufficient, if we say `the name is the first character of the service id`, the name is completely derived from the service id, but not unique

3. the service id is not the only valid option: if the service instance has other unique (combinations of fields), that are provided as unique (see previous section), they can also serve to make unique id's 


(point 2 and 3 can be considered too precise to bother)

sanderr replied:
2. I did take that into account but I see how my wording is quite ambiguous in that respect. I'll pay special attention to this.
3. I don't think I ever mention "service id" directly (I've tried to explicitly avoid that). I mostly speak generically of "identity", which corresponds to what you're saying. I'll see if I can better define what I mean by that somewhere.

sanderr replied:
2. I tried to improve this a bit.
3. Added a definition of what the document means when it says "identity".

wouterdb commented on docs/resource_sets.rst:
```diff
@@ -91,3 +93,168 @@ Limitations
+
+Generally, for every index on a set of attributes of an owned resource, at least one of the fields must be either derived from
+the identity of the service instance, or allocated in a safe manner by a plugin as described above. The same goes for every
+pair of resource id and agent. If the former constraint is not met, a full compile might fail, while if the latter is not met,
```
simplified english is better 
```suggestion
pair of resource id and agent. If the first constraint is not met, a full compile might fail, while if the second is not met,
```

wouterdb commented on docs/resource_sets.rst:
```diff
@@ -91,3 +93,168 @@ Limitations
+pair of resource id and agent. If the former constraint is not met, a full compile might fail, while if the latter is not met,
+the export will be rejected because two services are trying to configure the same resources.
+
+For example, consider the example model from before. If two networks with two routers each would be created, they would result
```
This confuses me, because routers tend to connect networks, they can never be contained in them. (interfaces of routers can be in networks)

sanderr replied:
I'll modify the example to use `Host` rather than `Router`.

sanderr commented on docs/resource_sets.rst:
```diff
@@ -91,3 +93,168 @@ Limitations
+
+TODO: check how trivial it would be to extend to the multi-service use case
+TODO: "On a high level" -> remove?
+TODO: should I simplify and just talk about resources rather than generic nodes? Less accurate but also less abstract
```
Still, this is a pretty big thing we'd be glancing over. I think that if I was reading documentation like this I'd like it to be covered. But perhaps I'm not the average end-user in that regard.

wouterdb replied:
In any case, we should structure it to make the main point understood, then refine to the details

sanderr commented on docs/resource_sets.rst:
```diff
@@ -91,3 +93,168 @@ Limitations
+TODO: "On a high level" -> remove?
+TODO: should I simplify and just talk about resources rather than generic nodes? Less accurate but also less abstract
+
+On a high level, to allow safe use of partial compiles, it is imperative that each service's refinements (through
```
Like the ones below, just move it up? Or would you like something more generic, not tied to the example?

wouterdb replied:
Moving it up might be sufficient, examples are good.

sanderr commented on docs/resource_sets.rst:
```diff
@@ -61,7 +61,7 @@ as the index example shows. This can result in undefinable behavior.
 Constraints and rules
 ************************
 
-Some restrictions and guidelines were implemented to guarantee that partial compilations function properly:
```
On the other hand, it might be easier to manage while the model is still small, then with each small change make sure to meet these constraints?

wouterdb commented on docs/resource_sets.rst:
```diff
@@ -91,3 +93,291 @@ Limitations
+    {"Resource(id=0)" "Resource(id=1)" "Resource(id=2)" "Resource(id=3)"} -> "SharedResource(id=0, value=0)" [style="invis"]
+
+
+.. digraph:: resource_sets_generic_bad
```
I would add the explanation of the bad one after the good one, this has a more natural flow.

wouterdb replied:
It may also help to split both bad things into their own example

sanderr replied:
Done

wouterdb commented on docs/resource_sets.rst:
```diff
@@ -91,3 +93,291 @@ Limitations
+Service instance uniqueness
+***************************
+With full compiles, indexes serve as the identity of a service instance in the model. The compiler then validates that no conflicting
+service instances can exist. With partial compiles this validation is lost because only one service instance will be present
```
```suggestion
service instances exist. With partial compiles this validation is lost because only one service instance will be present
```

sanderr replied:
Done

sanderr commented:
@wouterdb I finalized the document, review is for real this time. Changes since your last review:
- added warning to top of the document
- moved below model developer section
- added some links within the document
- improved on uniqueness and ownership section
- added testing section
- processed review comments
- updated PR description

sanderr commented on docs/model_developers/resource_sets.rst:
```diff
@@ -0,0 +1,478 @@
+
+    networks:
+        - id: 0
+          routers:
```
Oh thanks, they used to be called routers...

sanderr commented on docs/model_developers/resource_sets.rst:
```diff
@@ -0,0 +1,478 @@
+Service instance uniqueness
+---------------------------
+With full compiles, indexes serve as the identity of a service instance in the model. The compiler then validates that no conflicting
+service instances exist. With partial compiles this validation is lost because only one service instance will be present
```
It is possible but the document introduction states that it will not be covered here. Also lsm does not at the moment support it.

sanderr commented on docs/model_developers/resource_sets.rst:
```diff
@@ -0,0 +1,478 @@
+
+
+The ``read_from_yaml()`` plugin would have to verify that no two networks with the same id are defined. After this validation,
+if doing a partial, it may return a list with only the relevant network in it. For the yaml given above validation would fail
+because two routers with the same id are defined.
```
Double typo even: there are no routers, only hosts and networks, and you're right that it's about networks here.

sanderr commented on docs/model_developers/resource_sets.rst:
```diff
@@ -0,0 +1,478 @@
+
+Two arguments can be passed to the ``inmanta export`` command in order to export a partial model to the server:
+
+- ``--partial`` To specify that the model being compiled only contains the resources that need to be updated in relation to the previous version of the model.
+- ``--delete-resource-set <resource-set-name>`` This option, which may be used more than once, instructs the model to remove the resource set with the specified name. Only in conjunction with the preceding choice may this option be utilized. Note that utilizing a ``std::ResourceSet`` that includes no resources allows resource sets to be implicitly deleted during a partial compilation.
```
Good remark, I'll make sure to fix it here or create a ticket.

sanderr replied:
Turns out this is already specified in the "Resource sets" subsection. Do you think that suffices or do you have a suggestion to improve it?

sanderr commented on docs/model_developers/resource_sets.rst:
```diff
@@ -0,0 +1,478 @@
+
+The ``read_from_yaml()`` plugin would have to verify that no two networks with the same id are defined. After this validation,
+if doing a partial, it may return a list with only the relevant network in it. For the yaml given above validation would fail
+because two routers with the same id are defined.
```
```suggestion
because two networks with the same id are defined.
```

inmantaci commented:
Merged into branches master in a69cfa0ee4fc99a8b41db845b5c52991c6845ae0

inmantaci commented:
Processing #4891.
